dl/src/main/scala/com/intel/analytics/bigdl/nn/SoftmaxWithCriterion.scala

```diff
@@ -0,0 +1,166 @@
+ * this work for additional information regarding copyright ownership.
+ * Intel Corporation licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.NormMode.NormMode
+import com.intel.analytics.bigdl.nn.abstractnn.TensorCriterion
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * Computes the multinomial logistic loss for a one-of-many classification task,
+ * passing real-valued predictions through a softmax to get a probability distribution over classes.
+ * It should be preferred over separate SoftmaxLayer + MultinomialLogisticLossLayer
+ * as its gradient computation is more numerically stable.
+ * @param ignoreLabel (optional) Specify a label value that
+ *                    should be ignored when computing the loss.
+ * @param normalizeMode How to normalize the output loss.
+ */
+class SoftmaxWithCriterion[T: ClassTag](ignoreLabel: Option[Int] = None,
+  normalizeMode: NormMode = NormMode.VALID)
+  (implicit ev: TensorNumeric[T]) extends TensorCriterion[T] {
+  @transient var softmax: SoftMax[T] = _
+
+  @transient var prob: Tensor[T] = _
+
+  @transient var outerNum = 0 // batchsize
+
+  @transient var innerNum = 1
+
+  @transient var nClasses = 2
+
+  override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
+    // input: batchsize * softmaxAxis * height * width
+    // for this example, (1, 2, 90, 4) if we have 1 image batch, 2 classes, 9*10*4 anchors
+    outerNum = input.size(1)
+    innerNum = 1
+    var i = 3
+    while (i <= input.dim()) {
+      innerNum = innerNum * input.size(i)
+      i += 1
+    }
+    nClasses = input.size(2)
+    if (softmax == null) {
+      softmax = new SoftMax()
+    }
+    prob = softmax.forward(input)
+    val probData = prob.storage().array()
+    val labelData = target.storage().array()
+    var loss = ev.fromType(0)
+    val dim = prob.nElement() / outerNum
+    var count = 0
+    i = 0
+    while (i < outerNum) {
+      var j = 0
+      while (j < innerNum) {
+        val curTarget = ev.toType[Int](labelData(i * innerNum + j))
+        if (ignoreLabel.isEmpty || ignoreLabel.get != curTarget) {
+          assert(curTarget >= 0 && curTarget < nClasses,
+            s"curTarget $curTarget is out of range 0 to ${ nClasses - 1 } ")
+          loss = ev.minus(loss,
```

Comment:
please try to wrap this piece of code into a method and don't use generic inside.